src/mbgl/gl/gl_object_store.hpp

```diff
@@ -3,136 +3,75 @@
 #include <mbgl/gl/gl.hpp>
 #include <mbgl/util/noncopyable.hpp>
 
+#include <unique_resource.hpp>
+
 #include <array>
 #include <algorithm>
 #include <memory>
 #include <vector>
 
+#include <functional>
+
 namespace mbgl {
 namespace gl {
 
+constexpr GLsizei TextureMax = 64;
+
+using UniqueObject = std_experimental::unique_resource<GLuint, std::function<void(GLuint)>>;
```

Comment:
Rather than using `std::function` polymorphism, create a unique `*Deleter` functor and `Unique*` typedef for each resource type as sketched in https://github.com/mapbox/mapbox-gl-native/pull/5141#issuecomment-221719872. This adds additional type safety: mistaken use of the wrong resource type will be a compile time error.